compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeHashVisitor.java

```diff
@@ -0,0 +1,574 @@
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BAnyType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), type.getKind().typeName());
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BFutureType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), visit(type.constraint));
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BHandleType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), type.getKind().typeName());
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BMapType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), visit(type.constraint));
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BStreamType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), visit(type.constraint), visit(type.error));
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BTypedescType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), visit(type.constraint));
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BXMLType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), visit(type.constraint));
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BAnydataType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), type.getKind().typeName());
+        return addToVisited(type, hash);
+    }
+
+    @Override
+    public Integer visit(BErrorType type) {
+        if (isVisited(type)) {
+            return visited.get(type);
+        }
+        if (isCyclic(type)) {
+            return 0;
+        }
+        Integer hash = hash(baseHash(type), visit(type.detailType));
```

Comment:
@rdhananjaya Do we need to add type Id set as a precaution?

Reply:
Type id set hash included in error type hash and object type hash.